File: services/typescript/koa/src/routes/handlers/get.ts
import type { RouterContext } from "@koa/router";
import { type State, parseId } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function getHandler(state: State) {
  return (ctx: RouterContext) => {
    const logger = getLogger();

    const id = parseId(ctx.params.id);
    if (!id.success) {
      logger.warn({ method: ctx.req.method, path: ctx.path }, "Invalid path");
      return sendError(ctx, AppErrors.InvalidPath);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "GET", path: ctx.path },
        "Task not found"
      );
      return sendError(ctx, AppErrors.TaskNotFound);
    }

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        operation: task.operation.toString().toLowerCase(),
        method: "GET"
      },
      "Retrieved task"
    );

    ctx.status = 200;
    ctx.body = task;
  };
}

File: services/go/fiber/routes/handlers/get.go
package handlers

import (
	"app"
	"fiber/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gofiber/fiber/v3"
)

func GetHandler(state *app.AppState) fiber.Handler {
	return func(c fiber.Ctx) error {
		id, err := ParseID(c)
		if err != nil {
			return err
		}

		state.Mu.RLock()
		task, ok := state.Tasks[id]
		state.Mu.RUnlock()

		if !ok {
			slog.Warn("Task not found", "method", c.Method(), "path", c.Path(), "id", id.String())
			return routes.AppErrors.TaskNotFound.Error(c)
		}

		slog.Info(
			"Retrieved task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Method(),
		)

		return c.Status(http.StatusOK).JSON(task)
	}
}
